mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -3970,6 +3971,11 @@ struct FoldTensorCastProducerOp
     if (isa<InsertSliceOp>(op.getOperation()))
       return failure();
 
+    // scf::ForallOp also has its own folding logic taking the Types of its
+    // BlockArguments into consideration.
+    if (isa<scf::ForallOp>(op.getOperation()))
+      return failure();
+
```

Comment:
These lines are problematic as the tensor dialect does not depend on the SCF dialect.
To make this matter even worse, the SCF dialect currently depends on the Tensor dialect,, creating a circular dependency between the two. 
This leads to `BUILD_SHARED_LIBS` to fail building as well as any downstream user dependending on the tensor dialect to get linker errors that they need to include SCF as well now.

I sadly know very little about the tensor dialect/bufferzization/all that stuff, but is this kind of layering really necesseray?

Reply:
Oh right, I missed that. We should avoid such references to other dialects. `FoldTensorCastProducerOp` should not fold casts into destination style ops that have a region without further checks. At the moment there is no way to query whether such folding is safe or not.

Maybe we can do something with a new interface. We already have `CastOpInterface`. We could have another `FoldCastOpInterface` that ops can implement that support folding of cast ops into themselves.

Reply:
Thank you for pointing that out!

I just tried adding a simple check for `op->getNumRegions() != 0` but this also doesn't work because it excludes all of the `linalgOp`s that are some of the primary targets of this Pattern. How about excluding Ops that have tensor bbArgs, or is this getting too specific?

About the potential Interface - is there a way to implement this so that it doesn't add extra implementation effort to every DPS op?

Reply:
I think you can just look for `LoopLikeOpInterface` here... That adds a dependence from `DestPassingStyleOpInterface` to `LoopLikeOpInterface`, but we might be able to live with that.